### DISCUSSION TITLE
How to install OpenSSL 1.0.X FIPS?

### DISCUSSION THREAD

**Discussion by junaruga** [openssl/openssl#21904] [Q&A]
I am trying to install OpenSSL 1.0.2u (the latest version of the 1.0.2) with FIPS option. Sorry for a question for old version OpenSSL.

Here is my environment.

```
$ cat /etc/fedora-release 
Fedora release 38 (Thirty Eight)

$ gcc --version
gcc (GCC) 13.1.1 20230614 (Red Hat 13.1.1-4)
Copyright (C) 2023 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ pwd
/home/jaruga/git/openssl2

$ git remote -v | grep origin
origin	https://github.com/openssl/openssl.git (fetch)
origin	https://github.com/openssl/openssl.git (push)

$ git checkout OpenSSL_1_0_2u
```

I couldn't find a document about how to install FIPS in the OpenSSL_1_0_2u branch. Do you know where the document is? However, as I found my old issue ticket - https://github.com/openssl/openssl/issues/7582, I am trying the following way.

Then configure OpenSSL on the following way. Is this correct? Is the `--with-fipsdir` option required?

```
$ ./Configure shared fips \
   --prefix="$HOME/.local/openssl-1.0.2u" \
   --with-fipsdir="$HOME/.local/openssl-1.0.2u-fips" \
   linux-x86_64
...
Configured for linux-x86_64.

*** Because of configuration changes, you MUST do the following before
*** building:

  make depend

$ echo $?
0
```

Then running the `make depend` by following the message above, I saw the errors below. Do you know how to fix it?

```
$ make depend
making depend in crypto...
make[1]: Entering directory '/home/jaruga/var/git/openssl2/crypto'
In file included from /usr/include/openssl/fips.h:15,
                 from o_fips.c:61:
/usr/include/openssl/macros.h:155:4: error: #error "OPENSSL_API_COMPAT expresses an impossible API compatibility level"
  155 | #  error "OPENSSL_API_COMPAT expresses an impossible API compatibility level"
      |    ^~~~~
o_fips.c:62:11: fatal error: openssl/fips_rand.h: No such file or directory
   62 | # include <openssl/fips_rand.h>
      |           ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
In file included from /usr/include/openssl/fips.h:15,
                 from o_init.c:59:
/usr/include/openssl/macros.h:155:4: error: #error "OPENSSL_API_COMPAT expresses an impossible API compatibility level"
  155 | #  error "OPENSSL_API_COMPAT expresses an impossible API compatibility level"
      |    ^~~~~
In file included from /usr/include/openssl/fips.h:15,
                 from fips_err.h:64,
                 from fips_ers.c:4:
/usr/include/openssl/macros.h:155:4: error: #error "OPENSSL_API_COMPAT expresses an impossible API compatibility level"
  155 | #  error "OPENSSL_API_COMPAT expresses an impossible API compatibility level"
      |    ^~~~~
make[1]: *** [Makefile:139: local_depend] Error 1
make[1]: Leaving directory '/home/jaruga/var/git/openssl2/crypto'
make: *** [Makefile:483: depend] Error 1
```

**Comment by slontis** [comment]
In terms of Upstream OpenSSL FIPS validations see
https://csrc.nist.gov/projects/cryptographic-module-validation-program/validated-modules/search?SearchMode=Advanced&Vendor=OpenSSL&Standard=140-2&CertificateStatus=Historical&ValidationYear=0
You will see that these are all historical modules, which means that it is pointless to use the FOM.
You will see Security Policies associated in the above link which gives directions on how to build the FOM.

Assuming you are using a Redhat derived FIPS module you would need to find the corresponding Security Policy before using the code and matching the version in the documentation (which is probably not 1.0.2u). As there was a FIPS rule change that caused a whole lot of certificates to become invalid (SP80056A) for 1.0.2 validations it is not likely that you can use 1.0.2 for FIPS.

**Reply by junaruga** [reply]
Thanks for the answer. I understand the OpenSSL 1.0.X is already end of life.

Though my laptop is Fedora Linux, my intention to know how to install 1.0.X FIPS is that I want to add OpenSSL 1.0.2u FIPS case to [OpenSSL Ruby binding (Ruby's OpenSSL library](https://github.com/ruby/openssl)'s GitHub Actions CI (the host is Ubuntu). Ruby OpenSSL library still has an implementation for OpenSSL 1.0 FIPS. I am working for the project. The project has OpenSSL non-FIPS cases since OpenSSL 1.0.2u, and OpenSSL FIPS cases for 3.2, 3.1, 3.0 in the CI. As a reference, the GitHub Actions YAML config file is [here](https://github.com/ruby/openssl/blob/bdf060f85c5c297bb78c19d83c7511dbe7c418be/.github/workflows/test.yml#L73-L97).

**Reply by t8m** [reply]
Nobody should be using the old FOM for anything new. So I do not think there is a point in supporting it with new bindings.

**Comment by junaruga** [comment]
As a note for someone who is interested in this topic, I was able to build OpenSSL 1.0.2u on the OpenSSL_1_0_2u branch if I don't set `./Configure` FIPS options.

```
$ gcc --version
gcc (GCC) 13.2.1 20231011 (Red Hat 13.2.1-4)
Copyright (C) 2023 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ ./Configure \
  --prefix=$HOME/.local/openssl-1.0.2u-debug \
  --libdir=lib \
  shared \
  linux-x86_64 \
  '-Wl,-rpath,$(LIBRPATH)' \
  -O0 -g3 -ggdb3 -gdwarf-5

$ make -j4

$ make install

$ ~/.local/openssl-1.0.2u-debug/bin/openssl version
OpenSSL 1.0.2u  20 Dec 2019
```
